src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -372,18 +407,31 @@ void GcsTaskManager::HandleGetTaskEvents(rpc::GetTaskEventsRequest request,
         task_ids.insert(TaskID::FromBinary(task_filter_obj.task_id()));
       }
     }
-    if (!task_ids.empty()) {
+    if (task_ids.size() == 1) {
       task_events = task_event_storage_->GetTaskEvents(task_ids);
+    } else if (task_ids.size() > 1) {
+      task_events = std::vector<rpc::TaskEvents>();
     }
-  } else if (filters.has_job_filter() &&
-             filters.job_filter().predicate() == rpc::FilterPredicate::EQUAL) {
-    const auto job_id = JobID::FromBinary(filters.job_filter().job_id());
-    task_events = task_event_storage_->GetTaskEvents(job_id);
-    // Populate per-job data loss.
-    if (task_event_storage_->HasJob(job_id)) {
-      const auto &job_summary = task_event_storage_->GetJobTaskSummary(job_id);
-      reply->set_num_profile_task_events_dropped(job_summary.NumProfileEventsDropped());
-      reply->set_num_status_task_events_dropped(job_summary.NumTaskAttemptsDropped());
+  } else if (filters.job_filters_size() > 0) {
+    absl::flat_hash_set<JobID> job_ids;
+    for (const auto &job_filter_obj : filters.job_filters()) {
+      if (job_filter_obj.predicate() == rpc::FilterPredicate::EQUAL) {
+        job_ids.insert(JobID::FromBinary(job_filter_obj.job_id()));
+      }
+    }
+
+    if (job_ids.size() == 1) {
+      JobID job_id = *job_ids.begin();
+      task_events = task_event_storage_->GetTaskEvents(job_id);
+
+      // Populate per-job data loss.
+      if (task_event_storage_->HasJob(job_id)) {
+        const auto &job_summary = task_event_storage_->GetJobTaskSummary(job_id);
+        reply->set_num_profile_task_events_dropped(job_summary.NumProfileEventsDropped());
+        reply->set_num_status_task_events_dropped(job_summary.NumTaskAttemptsDropped());
+      }
+    } else if (job_ids.size() > 1) {
+      task_events = std::vector<rpc::TaskEvents>();
```

Comment:
just nit: i think for future readability, we could probably split this part from the routine below. Logically, there are two types of filters:
1. filter that we could use index: equal filter with indexes on them. 
2. filter that has to be applied per task (task event): all remaining filters.

Having a routine that split filters into `filters_with_index` and `per_task_filter` might help with both readability and performance I think (the perf gain is probably not too much) because we don't need to apply some per element on the selected tasks.